### PR TITLE
fix: テスト品質監査 — 全18件の再実装テスト修正 + 本番コード改善

### PR DESCRIPTION
**Pull request by ayutaz**
## Summary
- 全30テストファイルの品質監査を実施し、HIGH 7件 / MEDIUM 10件 / LOW 3件の問題を特定
- M0-M3の4フェーズ + レトロスペクティブ指摘対応で全18件を解消
- 681 passed, 0 failed (リベース後全テスト通過)

## 本番コード改善
- **8関数抽出**: `build_infer_forward`, `apply_ema_weights`, `apply_ema_shadow_params`, `apply_transfer_defaults`, `load_multispeaker_checkpoint`, `normalize_id_tensor`, `resolve_config_path`, `resolve_speaker_id`
- **重複排除**: `build_infer_forward` → `model.infer()` の薄いラッパーに (推論パイプライン重複を完全解消)
- **dead code削除**: `infer_onnx.py` DSP関数5件 (-96行)
- **API改善**: EMA I/O分離、`apply_transfer_defaults` 副作用削減、`normalize_id_tensor` public化

## テスト改善
- コピペ再実装テスト: **13件 → 0件**
- 壊れたテスト (AttributeError): **4件 → 0件**
- 無意味なテスト: **2件 → 削除済**
- `phonemize_japanese` 5ファイル重複 → conftest.py 共通化
- `mock_wavlm_discriminator` セッションスコープ共通fixture
- `@pytest.mark.unit` / `@pytest.mark.training` マーカー追加 (CI実行漏れ修正)
- `@pytest.mark.parametrize` テーブル駆動化 (M1-4, M1-6)
- `build_infer_forward` / `model.infer()` パリティテスト追加

## Test plan
- [x] `uv run pytest src/python/tests/ -m "unit"` — 681 passed, 3 skipped, 0 failed
- [x] 全M1-M3チケットの実装・レビュー・指摘対応完了
- [x] dev最新をリベースで取り込み済み (speaker_embedding対応含む)

### CODE REVIEW

File: src/python/piper_train/export_onnx.py [hunk 1]
```diff
@@ -25,6 +24,158 @@
...
+        A forward function with signature::
+
+            infer_forward(text, text_lengths, scales,
+                          sid=None, lid=None, prosody_features=None)
+            -> (audio: Tensor, durations: Tensor)
+    """
+
+    def infer_forward(
+        text, text_lengths, scales, sid=None, lid=None, prosody_features=None
+    ):
```
[the remark this answers is not in the document]
Comment (ayutaz) [reply src/python/piper_train/export_onnx.py:79]: 修正しました。build_infer_forward の返す関数に speaker_embedding/speaker_embedding_mask パラメータを追加し、model.infer() にパススルーするようにしました。model.infer() にも speaker_embedding_mask 対応を追加済みです。(321b23a9)

File: src/python/piper_train/export_onnx.py [hunk 2]
```diff
@@ -25,6 +24,158 @@
...
+        audio, _attn, _y_mask, _latents, durations = model.infer(
+            text,
+            text_lengths,
+            sid=sid,
+            lid=lid,
+            noise_scale=noise_scale,
+            length_scale=length_scale,
+            noise_scale_w=noise_scale_w,
+            prosody_features=prosody_features,
+        )
```
[the remark this answers is not in the document]
Comment (ayutaz) [reply src/python/piper_train/export_onnx.py]: 上記と同じコミットで対応済みです。(321b23a9)

File: src/python/piper_train/export_onnx.py [hunk 3]
```diff
@@ -25,6 +24,158 @@
...
+
+        model.onnx_export_mode = prev
```
[the remark this answers is not in the document]
Comment (ayutaz) [reply src/python/piper_train/export_onnx.py]: 修正しました。onnx_export_mode と dp.onnx_export_mode の両方を try/finally で保存・復元するようにしました。(321b23a9)

File: src/python/piper_train/export_onnx.py [hunk 4]
```diff
@@ -25,6 +24,158 @@
...
+    """
+    applied = 0
+    skipped = 0
+    dec_params = dict(decoder.named_parameters())
+    for name, shadow_param in shadow_params.items():
+        if name in dec_params:
+            dec_params[name].data.copy_(shadow_param)
+            applied += 1
+        else:
+            skipped += 1
```
[the remark this answers is not in the document]
Comment (ayutaz) [reply src/python/piper_train/export_onnx.py]: 修正しました。torch.no_grad() コンテキスト内で param.copy_(shadow_param) を使用する形に変更しました。(321b23a9)

File: src/python/piper_train/vits/models.py [hunk 5]
```diff
@@ -1100,7 +1102,7 @@ def infer(
         z = self.flow(z_p, y_mask, g=g, reverse=True)
         o = self.dec((z * y_mask)[:, :, :max_len], g=g)
 
-        return o, attn, y_mask, (z, z_p, m_p, logs_p)
+        return o, attn, y_mask, (z, z_p, m_p, logs_p), durations
```
[the remark this answers is not in the document]
Comment (ayutaz) [reply src/python/piper_train/vits/models.py]: InferOutput NamedTuple を導入しました。既存の tuple アンパック (audio, *_ = など) との互換性を維持しつつ、.durations で名前アクセスも可能です。全呼び出し元で動作確認済み。(4cb3299f)

File: src/python/tests/test_prosody_extraction.py [hunk 6]
```diff
@@ -512,63 +492,27 @@ def get_japanese_id_map():
...
-        # Valid data should have no errors
-        valid_errors = validate_dataset_prosody([json.dumps(d) for d in valid_data])
-        assert len(valid_errors) == 0, f"Valid data should have no errors: {valid_errors}"
-
-        # Invalid data should be detected
-        invalid_errors = validate_dataset_prosody([json.dumps(d) for d in invalid_data])
-        assert len(invalid_errors) == 1, f"Invalid data should have 1 error: {invalid_errors}"
-        assert invalid_errors[0]["phoneme_ids_len"] == 5
-        assert invalid_errors[0]["prosody_features_len"] == 3
+        """Test that prosody_features length mismatches are detectable in dataset entries."""
```
[the remark this answers is not in the document]
Comment (ayutaz) [reply src/python/tests/test_prosody_extraction.py]: ご指摘の通り、本番コードを呼び出していないため test_validate_dataset_prosody_lengths メソッドごと削除しました。(4cb3299f)

File: src/python/tests/test_prosody_extraction.py [hunk 7]
```diff
@@ -512,63 +492,27 @@ def get_japanese_id_map():
...
+            "phoneme_ids": [1, 2, 3, 4, 5],
+            "prosody_features": [
+                {"a1": 0, "a2": 0, "a3": 0},
+                {"a1": -1, "a2": 1, "a3": 3},
+                {"a1": 0, "a2": 2, "a3": 3},
+                {"a1": 1, "a2": 3, "a3": 3},
+                {"a1": 0, "a2": 0, "a3": 0},
+            ],
+        }
+        assert len(valid_entry["phoneme_ids"]) == len(valid_entry["prosody_features"])
```
[the remark this answers is not in the document]
Comment (ayutaz) [reply src/python/tests/test_prosody_extraction.py]: 上記と同じコミットで対応済みです。(4cb3299f)

File: src/python/tests/test_prosody_extraction.py [hunk 8]
```diff
@@ -512,63 +492,27 @@ def get_japanese_id_map():
...
+        # Invalid: 長さ不一致を検出できることを確認
+        invalid_entry = {
+            "phoneme_ids": [1, 2, 3, 4, 5],  # 5 items
+            "prosody_features": [
+                {"a1": 0, "a2": 0, "a3": 0},
+                {"a1": -1, "a2": 1, "a3": 3},
+                {"a1": 0, "a2": 2, "a3": 3},
+            ],  # 3 items — MISMATCH
+        }
+        assert len(invalid_entry["phoneme_ids"]) != len(invalid_entry["prosody_features"])
```
[the remark this answers is not in the document]
Comment (ayutaz) [reply src/python/tests/test_prosody_extraction.py]: 上記と同じコミットで対応済みです。(4cb3299f)

File: src/python/piper_train/__main__.py [hunk 9]
```diff
@@ -303,6 +303,109 @@ def create_parser():
...
+    """
+    _LOGGER.info("Resuming from multispeaker checkpoint: %s", checkpoint_path)
+
+    # 1. strict=False でロード（emb_g は自動スキップ）
+    # NOTE: weights_only=False is required to handle PosixPath objects in checkpoints
+    # This poses a security risk - only load trusted checkpoints
+    checkpoint = torch.load(
+        checkpoint_path,
+        map_location="cpu",
+        weights_only=False,
```
[the remark this answers is not in the document]
Comment (ayutaz) [reply src/python/piper_train/__main__.py:331]: この指摘は既存コード (本PRのスコープ外) に関するものです。torch.load の weights_only 対応は別途 issue で対応を検討します。
